### PR TITLE
Feat: Trace performance data

### PR DESCRIPTION
**Pull request by Fine0830**
Trace performance data
reference https://github.com/apache/skywalking/issues/4334

**screenshots**

屏幕快照 2020-08-06 下午5 05 04

### CODE REVIEW

File: src/performance/index.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    if (this.isPerf) {
+      this.perfConfig.performance = pagePerf.getPerfTiming();
+    }
+    const perfInfo = {
+      perfDetail: this.perfConfig.perfDetail,
+      resources: this.perfConfig.resources,
+      ...options,
+    };
+    new Report(options.reportUrl).sendByXhr(perfInfo);
+    this.clearPerf();
```
Comment (wu-sheng) [inline src/performance/index.ts]: What happens if the register hasn't completed?

Comment (Fine0830) [reply src/performance/index.ts]: I don't understand your question. The register of perf is `tracePerfDetail`. It doesn't seem to have the problem.

Comment (wu-sheng) [reply src/performance/index.ts]: I set the question for a wrong line, sorry. I mean for `#sendByXhr`. According to the protocol, https://github.com/apache/skywalking-data-collect-protocol/blob/master/browser/BrowserPerf.proto#L39-L41, the report data should include service id and version id. But I can't see the check and buffer(if not registered yet).

Comment (Fine0830) [reply src/performance/index.ts]: Oooh,  are these two parameters optional? If we don't pass ids, server seems to get data from service name and version id is current version. I forgot where it was explained. BTW, if we analysis data with time dimension, version doesn't seem to be need.

Comment (wu-sheng) [reply src/performance/index.ts]: No, those two parameters are required, that is why I asked. 
1. Version could be set as `-` by default, if the client-js doesn't set it, but this must notify the backend implementation.
1. Service Name is not existing in the protocol, only service ID, which means you have registered successfully. Do you want to add a name to the protocol as an alternative way?

My point is, this register would take time, you should have a buffer to hold this value until the register is done. Unless you should add name and version strings into the protocol. Also, you should know, the ID is required in the distributed tracing too. Even you add name and version in this protocol, you can't add those in the tracing protocol, the buffer mechanism is still required.

Comment (Fine0830) [reply src/performance/index.ts]: I update the PR, and I think we need service names to show in analysis UI.

**Comment by wu-sheng** [comment]
Does the PR support cache the performance data until the service and version register completed? If so, could you point out?

**Comment by wu-sheng** [comment]
> Does the PR support cache the performance data until the service and version register completed? If so, could you point out?

Don't need the register anymore. Please take a look at the latest protocol, if you have anything to update, please let me know.

**Comment by wu-sheng** [comment]
@arugal Please recheck the backend protocol too.

**Comment by Fine0830** [comment]
@arugal I added some metrics for performance detail
export type IPerfDetail = {
    redirectTime: number; // Time of redirection  
    dnsTime: number; // DNS query time  
    ttfbTime: number; // Time to First Byte  
    tcpTime: number; // tcp connection time  
    transTime: number; // Content transfer time  
    domAnalysisTime: number; // Dom parsing time  
    fptTime: number; // First Paint Time or Blank Screen Time  
    domReadyTime: number; // Dom ready time  
    loadPage: number; // Page full load time  
    resTime: number; // Synchronous load resources in the page  
    sslTime: number; // Only valid for HTTPS  
    ttlTime: number; // Time to interact  
    firstPackTime: number; // first pack time  
    fmpTime: number; // First Meaningful Paint  
};

**Comment by wu-sheng** [comment]
Please add comments for your codebase. We hope there will be more contributors when we have the 1st version.

**Comment by arugal** [comment]
> @arugal I added some metrics for performance detail
> export type IPerfDetail = {
> redirectTime: number; // Time of redirection
> dnsTime: number; // DNS query time
> ttfbTime: number; // Time to First Byte
> tcpTime: number; // tcp connection time
> transTime: number; // Content transfer time
> domAnalysisTime: number; // Dom parsing time
> fptTime: number; // First Paint Time or Blank Screen Time
> domReadyTime: number; // Dom ready time
> loadPage: number; // Page full load time
> resTime: number; // Synchronous load resources in the page
> sslTime: number; // Only valid for HTTPS
> ttlTime: number; // Time to interact
> firstPackTime: number; // first pack time
> fmpTime: number; // First Meaningful Paint
> };

Good to me :)  Need to update [BrowserPerf.proto](https://github.com/apache/skywalking-data-collect-protocol/blob/master/browser/BrowserPerf.proto).

**Comment by Fine0830** [comment]
Please recheck @wu-sheng @arugal

File: src/performance/type.d.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+export type IPerfDetail = {
+  redirectTime: number | undefined; // Time of redirection
+  dnsTime: number | undefined; // DNS query time
+  ttfbTime: number | undefined; // Time to First Byte
+  tcpTime: number | undefined; // Tcp connection time
+  transTime: number | undefined; // Content transfer time
+  domAnalysisTime: number | undefined; // Dom parsing time
+  fptTime: number | undefined; // First Paint Time or Blank Screen Time
+  domReadyTime: number | undefined; // Dom ready time
+  loadPage: number | undefined; // Page full load time
```
Comment (arugal) [inline src/performance/type.d.ts:37]: @Fine0830 In [BrowserPerf.proto](https://github.com/apache/skywalking-data-collect-protocol/blob/b2f381e63702bc43216ef5576637195102302c6e/browser/BrowserPerf.proto#L66), I changed `loadPage` to `loadPageTime`,

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**